ext4/dmverity/dmverity.go

```diff
@@ -0,0 +1,179 @@
+	HashBlockSize uint32
+	/* (72) The number of data blocks */
+	DataBlocks uint64
+	/* (80) Size of the salt */
+	SaltSize uint16
+	/* (82) Padding */
+	_ [6]byte
+	/* (88) The salt */
+	Salt [256]byte
+	/* (344) Padding */
+	_ [168]byte
+}
+
+// VerityInfo is minimal exported version of dmveritySuperblock
+type VerityInfo struct {
+	// Offset in blocks on hash device
+	HashOffsetInBlocks int64
+	// Set to true, when dm-verity super block is also written on the hash device
+	SuperBlock    bool
+	RootDigest    string
+	Salt          string
+	Algorithm     string
+	DataBlockSize uint32
+	HashBlockSize uint32
+	DataBlocks    uint64
+}
+
+// MerkleTree constructs dm-verity hash-tree for a given byte array with a fixed salt (0-byte) and algorithm (sha256).
+func MerkleTree(data []byte) []byte {
+	layers := make([][]byte, 0)
+
+	var currentLevel = bytes.NewBuffer(data)
+
+	for currentLevel.Len() != blockSize {
+		var blocks = currentLevel.Len() / blockSize
+		var nextLevel = bytes.NewBuffer(make([]byte, 0))
+
+		for i := 0; i < blocks; i++ {
+			block := make([]byte, blockSize)
+			_, _ = currentLevel.Read(block)
+			h := hash2(salt, block)
+			nextLevel.Write(h)
+		}
+
+		padding := bytes.Repeat([]byte{0}, blockSize-(nextLevel.Len()%blockSize))
+		nextLevel.Write(padding)
+
+		currentLevel = nextLevel
+		layers = append(layers, currentLevel.Bytes())
+	}
+
+	var tree = bytes.NewBuffer(make([]byte, 0))
+	for i := len(layers) - 1; i >= 0; i-- {
+		tree.Write(layers[i])
+	}
+
+	return tree.Bytes()
+}
+
+// RootHash computes root hash of dm-verity hash-tree
+func RootHash(tree []byte) []byte {
+	return hash2(salt, tree[:blockSize])
+}
+
+// NewDMVeritySuperblock returns a dm-verity superblock for a device with a given size, salt, algorithm and versions are
+// fixed.
+func NewDMVeritySuperblock(size uint64) *dmveritySuperblock {
+	superblock := &dmveritySuperblock{
+		Version:       1,
+		HashType:      1,
+		UUID:          generateUUID(),
+		DataBlockSize: blockSize,
+		HashBlockSize: blockSize,
+		DataBlocks:    size / blockSize,
+		SaltSize:      uint16(len(salt)),
+	}
+
+	copy(superblock.Signature[:], "verity")
+	copy(superblock.Algorithm[:], "sha256")
+	copy(superblock.Salt[:], salt)
+
+	return superblock
+}
+
+func hash2(a, b []byte) []byte {
+	h := sha256.New()
+	h.Write(append(a, b...))
+	return h.Sum(nil)
+}
+
+func generateUUID() [16]byte {
+	res := [16]byte{}
+	if _, err := rand.Read(res[:]); err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// ReadDMVerityInfo extracts dm-verity super block information and merkle tree root hash
+func ReadDMVerityInfo(vhdPath string, offsetInBytes int64) (*VerityInfo, error) {
+	vhd, err := os.OpenFile(vhdPath, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer vhd.Close()
+
+	// Skip the ext4 data to get to dm-verity super block
+	if s, err := vhd.Seek(offsetInBytes, io.SeekStart); err != nil || s != offsetInBytes {
+		return nil, errors.Wrap(err, "failed to seek dm-verity super block")
```

Comment:
if error is nil but `s != offsetInBytes` this error message wouldn't be very helpful

Reply:
agreed